src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonRequestHelpers.cs

```diff
@@ -173,66 +174,88 @@ public static async Task WriteResponseMessage(HttpResponse response, Encoding en
-                    }
-                    else
+                    bodyContent = httpBody;
+                }
+                else
+                {
+                    if (!serverCallContext.IsJsonRequestContent)
                     {
-                        IMessage bodyContent;
+                        GrpcServerLog.UnsupportedRequestContentType(serverCallContext.Logger, serverCallContext.HttpContext.Request.ContentType);
+                        throw new RpcException(new Status(StatusCode.InvalidArgument, "Request content-type of application/json is required."));
+                    }
 
-                        try
-                        {
-                            GrpcServerLog.DeserializingMessage(serverCallContext.Logger, serverCallContext.DescriptorInfo.BodyDescriptor.ClrType);
-                            bodyContent = (IMessage)(await JsonSerializer.DeserializeAsync(stream, serverCallContext.DescriptorInfo.BodyDescriptor.ClrType, serializerOptions))!;
-                        }
-                        catch (JsonException)
-                        {
-                            throw new RpcException(new Status(StatusCode.InvalidArgument, "Request JSON payload is not correctly formatted."));
-                        }
-                        catch (Exception exception)
-                        {
-                            throw new RpcException(new Status(StatusCode.InvalidArgument, exception.Message));
-                        }
+                    var (stream, usesTranscodingStream) = GetStream(serverCallContext.HttpContext.Request.Body, serverCallContext.RequestEncoding);
 
-                        if (serverCallContext.DescriptorInfo.BodyFieldDescriptors != null)
+                    try
+                    {
+                        if (serverCallContext.DescriptorInfo.BodyDescriptorRepeated)
                         {
                             requestMessage = (IMessage)Activator.CreateInstance<TRequest>();
-                            ServiceDescriptorHelpers.RecursiveSetValue(requestMessage, serverCallContext.DescriptorInfo.BodyFieldDescriptors, bodyContent!); // TODO - check nullability
+
+                            // TODO: JsonSerializer currently doesn't support deserializing values onto an existing object or collection.
+                            // Either update this to use new functionality in JsonSerializer or improve work-around perf.
+                            var type = JsonConverterHelper.GetFieldType(serverCallContext.DescriptorInfo.BodyFieldDescriptors.Last());
+                            var listType = typeof(List<>).MakeGenericType(type);
+
+                            GrpcServerLog.DeserializingMessage(serverCallContext.Logger, listType);
+
+                            bodyContent = (await JsonSerializer.DeserializeAsync(stream, listType, serializerOptions))!;
```

Comment:
Does this need the same `try catch` as below?

Reply:
Yes, it probably should